### DISCUSSION TITLE
YouValidateMe JSON Schema library server project built on this library

### DISCUSSION THREAD

**Discussion by crowdwave** [santhosh-tekuri/jsonschema#169] [Show and tell]
Hello I thought you might be interested, I built a JSON schema library server and it is using your library for its core.
 
https://github.com/crowdwave/youvalidateme

I started with a different JSON Schema library for Go but I switched to yours as your has more features and seems more actively maintained.

Thanks!

**Comment by santhosh-tekuri** [comment]
@crowdwave happy to hear that

you might want to try `v6.0.0-beta1` from https://github.com/santhosh-tekuri/jsonschema/tree/boon

this has some new features. note that api is slightly different since it is major version bump

**Reply by crowdwave** [reply]
Thanks.

Yes I'm having some trouble getting schemas to valid against any of the drafts is there documentation for the new syntax?

What I would really appreciate is this example extended further to show how to validate schemas against the various Draft specs and also how to Compliler.LoadURL and validate.

https://github.com/santhosh-tekuri/jsonschema/blob/boon/example_regexp_test.go

All I need to do is use version 6 with the ECMA regex plus, valid schemas against drafts and valid JSON against schemas.

Is there an example that ties all that together into some example code?

thanks!

**Comment by santhosh-tekuri** [comment]
as per specification `$schema` keyword specifies which draft to be used for the schema.

if the schema does not has `$schema` then the library is free to chose whatever draft it wants to. Currently this library default to latest draft in such cases. In case you want to force to use draft6 you can do `compiler.DefaultDraft(jsonschema.Draft6)`

**Comment by santhosh-tekuri** [comment]
in your project you run the server:

```
./youvalidateme --port 8080 --schemas-dir=/path/to/schemas
```

so you want to read all schema files only from the path specified by user and you don't want to expose any system file path to user. with `6.0.0-beta1` you can do it as follows:

we want to introduce custom url for this; for example `schemas:///dir1/dir2/test.json` will translate to `/path/to/schemas/dir1/dir2/test.json`
```go
type dirLoader string

func (l dirLoader) Load(url string) (any, string) {
    path, ok := strings.CutPrefix(url, "schemas:///")
    if !ok {
        return nil, fmt.Errorf("unsupported url: %v", url)
    }
    path = filepath.join(string(l), path)
    f, err := os.Open(path)
    if err!=nil {
        return nil, err
    }
    defer f.Close()
    return jsonschema.UnmarshalJSON(f)
}

func testLoader() {
       c := jsonschema.NewCompiler()
       c.DefaultDraft(jsonschema.Draft6)
       c.useLoader(dirLoader("/path/to/schemas"))

      // to compile a particular schema
      c.Compile("schemas:///dir1/dir2/test.json")
}
```
